### ISSUE TITLE
Implement peer-discovery interface

### ISSUE THREAD

**Issue by lgierth** [libp2p/go-libp2p#222]
As per https://github.com/libp2p/interface-peer-discovery

This interface would be implemented by:
- mdns
- bootstrap
- random walk (does this currently exist in go-libp2p?)
- in the future: webrtc, ...

**Comment by vyzo** [comment]
do we want this to be dht based?
It could be the solution to the generalized rendezvous problem, and be usable for pubsub, relay, etc

**Comment by Kubuxu** [comment]
Additional provider of peer-discovery is "old peers". This way we could make the network independent from the bootstrap nodes after first run.

**Comment by diasdavid** [comment]
> do we want this to be dht based? 

DHT is Peer Routing + Content Routing. Read:
- https://github.com/libp2p/js-libp2p/tree/master/examples/peer-and-content-routing
- https://github.com/libp2p/specs/blob/master/4-architecture.md#41-peer-routing
- https://github.com/libp2p/specs/blob/master/4-architecture.md#43-distributed-record-store
- https://github.com/ipfs/specs/issues/1

Peer Discovery is things MulticastDNS or Bootstrap. Read:
- https://github.com/libp2p/js-libp2p/tree/master/examples/discovery-mechanisms
- https://github.com/libp2p/specs/blob/master/4-architecture.md#44-discovery

> peer-discovery is "old peers"

Yeah, we can have a resurrect Peer Discovery that knows how to store to disk old peers and then brings them back on node boot. 👍 👍

**Comment by deckbsd** [comment]
Hello everybody,

If i understand correctly, does that mean that actually from a go-libp2p peer i can't try to find some other peers ? that's why i was wondering how, when i start a host i can't make that host to connect to any other nodes. Because actually in the examples, each nodes are created like in client/server basic model (the client know the ip of the "server"(an other node that it can communicate with). m i wrong ?

**Comment by Kubuxu** [comment]
> If i understand correctly, does that mean that actually from a go-libp2p peer i can't try to find some other peers ?

No, find peers works (it is sometimes difficult in case of complicated/layered NATs).

This issue is about internal go-libp2p interface allowing better code structure when it comes to peer discovery.

**Comment by agahEbrahimi** [comment]
@Kubuxu Are there any examples of the peer discovery? I've been wandering around for 2-3 days trying to find an example. If not, can anyone give me a short sample? 
Thank you

**Comment by Kubuxu** [comment]
@agahEbrahimi most of them should work by default. We are working on clearing up the bootstraping but for an example you can look here: https://github.com/libp2p/go-libp2p/pull/302

Random walk (DHT) should work by default and so should mdns AFAIK.

**Comment by daviddias** [comment]
Can someone point me to an up to date example of how to enable/disable random-walk in go-libp2p?

**Comment by daviddias** [comment]
Just did a quick "chase the PR" exercised and discovered that:

- There is an example created for Bootstrapping https://github.com/libp2p/go-libp2p/pull/302
- There is a refactor by @florianlenz to make go-libp2p-bootstrap nicer https://github.com/florianlenz/go-libp2p-bootstrap/pull/3

Still looking for random-walk

**Comment by daviddias** [comment]
@aschmahmann I would say no but given that there is a bunch of terms that are overloaded, the best is for me to clarify my intent: What I want to achieve is to enable/disable the routine that searches the DHT for a random peer (most likely that doesn't exist) with the purpose of refreshing the routing tables. Is there a way to enable/disable this (known in the P2P literature as random-walk)?

**Comment by yusefnapora** [comment]
@daviddias I think from reading through the [kad bootstrap code](https://github.com/libp2p/go-libp2p-kad-dht/blob/4fd64988b55d6e93405eceffeb3037da46d1848b/dht_bootstrap.go) a bit, that "bootstrap" in the context of the libp2p kad DHT does mean random-walk. 

It looks like by default it does a query for a random peer id in each bucket that hasn't been refreshed within some interval, plus a periodic query for our own id.

If you set the `DisableAutoBootstrap` option that @aschmahmann mentioned, it looks like it won't do the bootstrapping on a periodic timer, but it would still be possible to manually trigger the bootstrap process by calling the `Bootstrap` method.

**Comment by daviddias** [comment]
> @daviddias I think from reading through the kad bootstrap code a bit, that "bootstrap" in the context of the libp2p kad DHT does mean random-walk.

Aaaaaaah!!! Ok, so in go-libp2p bootstrap gets used for:

- Connecting to set of bootstrap nodes when the node boots up (aka railing into the network in p2p literature)
- Refreshing the Kademlia Routing Tree every now and then (aka random-walk in the p2p literature)

> It looks like by default it does a query for a random peer id in each bucket that hasn't been refreshed within some interval, plus a periodic query for our own id.

How does it actually achieve this? Given that we sha2-256 the peerIds for the Kad Routing Table in our DHT implementation, one would actually have to brute force to find another PeerId that would fall within a specific k-bucket.

Anyhow, given that random-walking is walking randomly over new ways, most k-buckets will get somewhat refreshed along the way.

> If you set the DisableAutoBootstrap option that @aschmahmann mentioned, it looks like it won't do the bootstrapping on a periodic timer, but it would still be possible to manually trigger the bootstrap process by calling the Bootstrap method.

Thank you! This is exactly what I needed. ❤️ @aschmahmann @yusefnapora

**Comment by aschmahmann** [comment]
> How does it actually achieve this? Given that we sha2-256 the peerIds for the Kad Routing Table in our DHT implementation, one would actually have to brute force to find another PeerId that would fall within a specific k-bucket.

Sooo.... funny story about that we basically pre-brute force 1 per bucket for the first 16 buckets.

Precomputed buckets: https://github.com/libp2p/go-libp2p-kbucket/blob/8b77351e0f784a5f71749d23000897c8aee71a76/bucket_prefixmap.go#L4

Reasoning: https://github.com/libp2p/go-libp2p-kbucket/pull/38/files#r309983750

**Comment by daviddias** [comment]
# 🤯

Clever though!

**Comment by daviddias** [comment]
@aschmahmann @yusefnapora your tips worked like a charm https://github.com/ipfs/testground/pull/116/commits/c874df088c312f9e3871c51e7cd35881ef6aa3c9

**Comment by aarshkshah1992** [comment]
For posterity:

@aschmahmann's suggestion is bang on. Please refer to this [writeup](https://discuss.libp2p.io/t/go-libp2p-kad-dht-0-3-0-released-with-breaking-changes/337) for a good explanation of how "random walks"/"Dht refresh" works & how to use this feature.

**Comment by melroy89** [comment]
These very old issues from 2019 can be closed, right? Are these all still relevant?
